[[ Bug 16434 ]] Fix array operations in 7.0

livecodeali opened the pull request:
This patch amends the array set operations so that they have simple
and well-defined semantics-
Union:

```
 repeat for each key tKey in tRight
    if tKey is not among the keys of tLeft then
        put tRight[tKey] into tLeft[tKey]
    else if tRecursive then
        union tLeft[tKey] with tRight[tKey] recursively
    end if
 end repeat
```

Intersect:

```
 repeat for each key tKey in tLeft
    if tKey is not among the keys of tRight then
        delete variable tLeft[tKey]
    else if tRecursive then
        intersect tLeft[tKey] with tRight[tKey] recursively
    end if
 end repeat
```

This is a change from 6.7, in that the recursive behavior is now consistent
with the top-level. This affects only the following two invocations of array
intersect:

`intersect tLeftString with tRightString`
`intersect tLeftString with tRightArray`

both of which result in tLeftString being empty in 6.7. Now these both result
in tLeftString being unchanged.

Commit:
[[ Bug 16434 ]] Fix array operations in 7.0

peter-b commented:
@livecodeali I've already created `tests/lcs/core/array/` for array tests.  Is there any chance I could persuade you to move your new test file from the `arrays` directory to the `array` directory, please?

livecodeali commented:
@peter-b Sure! This is WIP after all :-)

livecodeali committed:
[[ Array Set Ops ]] Fix semantics of intersect and union

This patch amends the array set operations so that they have simple
and well-defined semantics-
Union:
```
 repeat for each key tKey in tRight
    if tKey is not among the keys of tLeft then
        put tRight[tKey] into tLeft[tKey]
    else if tRecursive then
        union tLeft[tKey] with tRight[tKey] recursively
    end if
 end repeat
```

Intersect:
```
 repeat for each key tKey in tLeft
    if tKey is not among the keys of tRight then
        delete variable tLeft[tKey]
    else if tRecursive then
        intersect tLeft[tKey] with tRight[tKey] recursively
    end if
 end repeat
```

This is a change from 6.7, in that the recursive behavior is now consistent
with the top-level. This affects only the following two invocations of array
intersect:

` intersect tLeftString with tRightString`
` intersect tLeftString with tRightArray`

both of which result in tLeftString being empty in 6.7. Now these both result
in tLeftString being unchanged.

livecodeali commented:
Argh, targeted at the wrong branch anyway!
